# Lab 4: Preparing Data for a Random Forest Model

Date of submission: 2020-12-11

Group-3 & members:
- Name (ID): Abhinav Sharma (0753639)-sec:006
- Name (ID): Nishant Mahajan (0753638)-sec:001
- Name (ID): Prashanth Nuthula (0753653)-sec:005
- Name (ID): Sai Sahas Elluru (0753808)-sec:003
- Name (ID): Siddharth Singh (0756590)-sec:005

The main goal of this lab is to prepare a dataset so that it can be used to train a model. 

**Instructions:**
- use the dataset assigned to your group;
- all features should end up being numeric;
- all missing values should be handled in a reasonable manner;
- as many features as possible should be retained (that is, you should have a valid reason to drop a feature); 
- some sort of feature engineering OR feature extraction should be attempted;
- NO model needs to be trained and evaluated. 

**Notes:**
- information about the data can be found at [this website](https://www.fueleconomy.gov/feg/download.shtml)
- the target variable is: 
    - `Comb Unadj FE - Conventional Fuel`
- you may remove any other column that is related to fuel economy; look for things like `City`, `Hwy`, `MPG`, etc;
- you may assume that we would be using a random forest model to train on the data;
- while you won't be able to make decisions about which features are or are not important based on output from a model, all decisions about how you prepare the data should be logical and valid for testing with a model;
- the following course notebooks should act as a guide:
    - *Part 1 - Categorical Variables*
    - *Part 2 - Categorical Variables*
    - *Part 3 - Categorical Variables*
    - *Part 4 - Categorical Variables*
    - *Part 1 - Missing Data*
    - *Part 2 - Missing Data*
    - *Final Pass*

**Grading:** 
- 50% of the grade will come from error-free code that encompasses all the criteria outlined in the instructions and written in Python/Pandas. 
- 50% will come from documenting what was done to the data and why it was done.  



**What to submit:**
- A copy of this notebook with:
    - Error-free code in Python/Pandas
    - All code cells executed and output visible



### Part 1: Preparing Data for a Model

In this part, you will provide the **final** code used to prepare the data for use with a random forest model. That is, the final output should be a dataframe with all features as a numeric data type and no missing values. 

To help in viewing the data, you may find these commands helpful (see [here](https://thispointer.com/python-pandas-how-to-display-full-dataframe-i-e-print-all-rows-columns-without-truncation/) for explanation):
- `pd.set_option('display.max_rows', None)`
- `pd.set_option('display.max_columns', None)`
- `pd.set_option('display.width', None)`
- `pd.set_option('display.max_colwidth', -1)`

In [213]:
import numpy as np
import pandas as pd
import os
#os.chdir('E:\\Clair\\SEM3\\ML\\Lab4')
data_1 = pd.read_csv('vehicles_2018.csv')
data= pd.read_csv('vehicles_2018.csv')

In [214]:
#Head of the vehicles 2018 data
data.head()

Unnamed: 0  Model Year      Mfr Name                  Division  \
0  0           2018        aston martin  Aston Martin Lagonda Ltd   
1  1           2018        aston martin  Aston Martin Lagonda Ltd   
2  2           2018        aston martin  Aston Martin Lagonda Ltd   
3  3           2018        aston martin  Aston Martin Lagonda Ltd   
4  4           2018        BMW           Mini                       

                  Carline Verify Mfr Cd  Index (Model Type Index)  Eng Displ  \
0  DB11 V12                ASX           1                         5.2         
1  DB11 V8                 ASX           2                         4.0         
2  Vanquish, S and Zagato  ASX           4                         6.0         
3  Rapide S                ASX           5                         6.0         
4  COOPER CLUBMAN          BMX           30                        1.5         

   # Cyl Transmission  City FE (Guide) - Conventional Fuel  \
0  12     Auto(S8)     15                                    
1  8      Auto(S8)     18                                    
2  12     Auto(S8)     13                                    
3  12     Auto(S8)     14                                    
4  3      Auto(S6)     24                                    

   Hwy FE (Guide) - Conventional Fuel  Comb FE (Guide) - Conventional Fuel  \
0  21                                  17                                    
1  24                                  20                                    
2  21                                  16                                    
3  21                                  17                                    
4  32                                  27                                    

   City Unadj FE - Conventional Fuel  Hwy Unadj FE - Conventional Fuel  \
0  18.8306                            30.8924                            
1  22.7207                            33.8295                            
2  16.5327                            28.7057                            
3  17.2984                            29.8618                            
4  31.3604                            46.5690                            

   Comb Unadj FE - Conventional Fuel  City Unrd Adj FE - Conventional Fuel  \
0  22.8444                            15.1397                                
1  26.6603                            18.1323                                
2  20.4316                            13.4660                                
3  21.3382                            14.0538                                
4  36.7632                            24.3407                                

   Hwy Unrd Adj FE - Conventional Fuel  Comb Unrd Adj FE - Conventional Fuel  \
0  20.5453                              17.1729                                
1  24.1217                              20.4132                                
2  20.7097                              15.9815                                
3  21.4866                              16.6449                                
4  32.2701                              27.3667                                

  Guzzler?  Air Aspir Method Air Aspiration Method Desc Trans      Trans Desc  \
0  NaN       TC               Turbocharged               SA    Semi-Automatic   
1  NaN       TC               Turbocharged               SA    Semi-Automatic   
2  G         NaN              Naturally Aspirated        SA    Semi-Automatic   
3  G         NaN              Naturally Aspirated        SA    Semi-Automatic   
4  NaN       TC               Turbocharged               SA    Semi-Automatic   

   Trans, Other  # Gears Lockup Torque Converter Trans Creeper Gear Drive Sys  \
0 NaN            8        Y                       N                  R          
1 NaN            8        Y                       N                  R          
2 NaN            8        Y                       N                  R          
3 NaN            8        Y                       N        

In [215]:
# The below code is from class notebooks part-1 missing_data
def sniff_modified(df):
    with pd.option_context("display.max_colwidth", 20):
        info = pd.DataFrame()
        info['data type'] = df.dtypes
        info['percent missing'] = df.isnull().sum()*100/len(df)
        info['No. unique'] = df.apply(lambda x: len(x.unique()))
        info['unique values'] = df.apply(lambda x: x.unique())
        return info.sort_values('data type')

In [216]:
# This function shows data type, missing values percentage, number of unique values and unique values of all columns
sniff_modified(data)

data type  \
Unnamed: 0                                                                                                 int64      
Carline Class                                                                                              int64      
EPA FE Label Dataset ID                                                                                    int64      
Annual Fuel1 Cost - Conventional Fuel                                                                      int64      
FE Rating (1-10 rating on Label)                                                                           int64      
GHG Rating (1-10 rating on Label)                                                                          int64      
Exhaust Valves Per Cyl                                                                                     int64      
#1 Mfr Smog Rating (Mfr Smog 1-10 Rating on Label for Test Group 1)                                        int64      
# Gears                                                                                                    int64      
City CO2 Rounded Adjusted                                                                                  int64      
Hwy CO2 Rounded Adjusted                                                                                   int64      
Comb FE (Guide) - Conventional Fuel                                                                        int64      
Intake Valves Per Cyl                                                                                      int64      
City FE (Guide) - Conventional Fuel                                                                        int64      
Comb CO2 Rounded Adjusted (as shown on FE Label)                                                           int64      
Model Year                                                                                                 int64      
Index (Model Type Index)                                                                                   int64      
# Cyl                                                                                                      int64      
Hwy FE (Guide) - Conventional Fuel                                                                         int64      
Charge Depleting Calc Appr Desc (PHEV only)                                                                float64    
Charge Depleting Calc Appr Code (PHEV only)                                                                float64    
Fuel Metering Type 2 Desc                                                                                  float64    
PHEV Total Driving Range (rounded to nearest 10 miles)DISTANCE                                             float64    
Fuel Metering Type 1 Desc                                                                                  float64    
# Drive Motor Gen                                                                                          float64    
Fuel Cell Onboard H2 Capacity (kg)                                                                         float64    
Usable H2 Fill Capacity (kg)                                                                               float64    
Fuel Cell Desc                                                                                             float64    
Comments                                                                                                   float64    
# Capacitors                                                                                               float64    
Batt Specific Energy (Watt-hr/kg)                                                                          float64    
Batt Energy Capacity (Amp-hrs)                                                                             float64    
Total Voltage for Battery Pack(s)                                                                          float64    
Battery Type, If Other                                      

In [217]:
# This code is to display all the rows and columns
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)
pd.set_option('display.max_colwidth', -1)

C:\Users\saisa\Anaconda3\envs\DAB-300\lib\site-packages\ipykernel_launcher.py:5: FutureWarning: Passing a negative integer is deprecated in version 1.0 and will not be supported in future version. Instead, use None to not limit the column width.
  """


In [218]:
# Removing columns where nulls are 100%
info = pd.DataFrame()
info['percent missing'] = data.isnull().sum()*100/len(data)
null = info.loc[info['percent missing'] == 100]


In [219]:
null_i = list(null.index)
for i in range(len(null_i)):
    data.drop(null_i[i],axis=1,inplace=True)
data.shape

(1290, 124)

In [220]:
pd.set_option('max_rows',163)
data.isnull().sum()/len(data)*100

Unnamed: 0                                                                                                   0.000000 
Model Year                                                                                                   0.000000 
Mfr Name                                                                                                     0.000000 
Division                                                                                                     0.000000 
Carline                                                                                                      0.000000 
Verify Mfr Cd                                                                                                0.000000 
Index (Model Type Index)                                                                                     0.000000 
Eng Displ                                                                                                    0.000000 
# Cyl                                           

In [221]:
# Dropping the columns which are related to fuel economy as per instructions
data.drop(columns=['City FE (Guide) - Conventional Fuel', 'Hwy FE (Guide) - Conventional Fuel', 'Comb FE (Guide) - Conventional Fuel', 'City Unadj FE - Conventional Fuel', 'Hwy Unadj FE - Conventional Fuel', 'City Unrd Adj FE - Conventional Fuel', 'Hwy Unrd Adj FE - Conventional Fuel', 'Comb Unrd Adj FE - Conventional Fuel', 'City2 FE (Guide) - Alternative Fuel', 'Hwy2 Fuel FE (Guide) - Alternative Fuel', 'Comb2 Fuel FE (Guide) - Alternative Fuel', 'City2 Unadj FE - Alternative Fuel', 'Hwy2 Unadj FE - Alternative Fuel', 'Comb2 Unadj FE - Alternative Fuel', 'City2 Unrd Adj FE - Alternative Fuel', 'Hwy2 Unrd Adj FE - Alternative Fuel', 'Cmb2 Unrd Adj FE - Alternative Fuel', 'EPA FE Label Dataset ID', 'Comb CO2 Rounded Adjusted (as shown on FE Label)'], axis=1, inplace=True)

In [222]:
data.shape

(1290, 105)

In [223]:
data.isnull().sum()/len(data)*100

Unnamed: 0                                                                                                   0.000000 
Model Year                                                                                                   0.000000 
Mfr Name                                                                                                     0.000000 
Division                                                                                                     0.000000 
Carline                                                                                                      0.000000 
Verify Mfr Cd                                                                                                0.000000 
Index (Model Type Index)                                                                                     0.000000 
Eng Displ                                                                                                    0.000000 
# Cyl                                           

In [224]:
# Dropping all the description columns
data.drop(data.columns[data.columns.str.contains('Desc',case=False)], axis=1,inplace=True)

In [225]:
# Dropping all passenger volume related columns
data.drop(data.columns[data.columns.str.contains('Pass',case=False)], axis=1,inplace=True)

In [226]:
# Dropping all luggage volume related columns
data.drop(data.columns[data.columns.str.contains('Lugg',case=False)], axis=1,inplace=True)

In [227]:
# dropping all battery realted columns
data.drop(data.columns[data.columns.str.contains('Batt',case=False)], axis=1,inplace=True)

In [228]:
data.drop(data.columns[data.columns.str.contains('Regen',case=False)], axis=1,inplace=True)

In [229]:
# dropping all fuel 2 related columns
data.drop(data.columns[data.columns.str.contains('Fuel2',case=False)], axis=1,inplace=True)

In [230]:
# Dropping comments
data.drop(data.columns[data.columns.str.contains('Comments',case=False)], axis=1,inplace=True)

In [231]:
# Dropping all the columns which cannot be imputed due to high null values
data.drop(columns=['Guzzler? ','Range1 - Model Type Driving Range - Conventional Fuel','Fuel Unit - Conventional Fuel',' Range2 - Alt Fuel Model Typ Driving Range - Alternative Fuel','Relabel','# Drive Motor Gen','Motor Gen Type, If Other','Rated Motor Gen Power (kW)','Fuel Cell Vehicle (Y or N)','Off Board Charge Capable (Y or N)','GHG 1-10 rating on Ethanol (EPA Determined)','#2 Smog Rating Test Group','#2 Mfr Smog Rating (Mfr Smog 1-10 Rating on Label for Test Group 2)','Unnamed: 0'],axis=1,inplace=True)

In [232]:
# Dropping all the columns which are unrelated or single value columns
data.drop(columns=['Transmission','Mfr Name','Release Date','Model Year','Suppressed?','Police/Emerg?'],axis=1,inplace=True)

In [233]:
data['Calc Approach Desc'] = data_1['Calc Approach Desc'].values

In [234]:
data.isnull().sum()/len(data)*100

Division                                                                                                     0.000000 
Carline                                                                                                      0.000000 
Verify Mfr Cd                                                                                                0.000000 
Index (Model Type Index)                                                                                     0.000000 
Eng Displ                                                                                                    0.000000 
# Cyl                                                                                                        0.000000 
Comb Unadj FE - Conventional Fuel                                                                            0.000000 
Air Aspir Method                                                                                             43.023256
Trans                                           

In [235]:
# Imputing the values
data["Air Aspir Method"].fillna("NAT", inplace = True)

In [236]:
print(data[data['Max Ethanol % - Gasoline'].isnull()])

       Division                        Carline Verify Mfr Cd  \
69   BMW         328d                           BMX            
70   BMW         328d xDrive                    BMX            
71   BMW         328d xDrive Sports Wagon       BMX            
103  BMW         540d xDrive                    BMX            
106  BMW         X5 xDrive 35d                  BMX            
188  Jeep        Grand Cherokee 4X2             CRX            
190  Jeep        Grand Cherokee 4X4             CRX            
211  RAM         1500 4X2                       CRX            
215  RAM         1500 4X4                       CRX            
366  Ford        F150 PICKUP 2WD                FMX            
367  Ford        F150 PICKUP 4WD                FMX            
370  Ford        F150 2WD BASE PAYLOAD LT TIRE  FMX            
371  Ford        F150 PICKUP 4WD XL/XLT         FMX            
372  Ford        F150 4WD BASE PAYLOAD LT TIRE  FMX            
392  Chevrolet   EQUINOX FWD            

In [237]:
# Relacing nulls with 0
data["Max Ethanol % - Gasoline"].fillna(0, inplace = True)

In [238]:
# Replacing nulls with 0
data["Max Biodiesel %"].fillna(0, inplace = True)

In [239]:
# Extracting the new feature using 2 columns and dropping the old columns
data["$ You Save over 5 years (amount saved in fuel costs over 5 years - on label) "].fillna(0, inplace = True)

In [240]:
print(data[data['$ You Save over 5 years (amount saved in fuel costs over 5 years - on label) '].isnull()])

Empty DataFrame
Columns: [Division, Carline, Verify Mfr Cd, Index (Model Type Index), Eng Displ, # Cyl, Comb Unadj FE - Conventional Fuel, Air Aspir Method, Trans, # Gears, Lockup Torque Converter, Trans Creeper Gear, Drive Sys, Max Ethanol % - Gasoline, Max Biodiesel %, Fuel Usage  - Conventional Fuel, Gas Guzzler Exempt (Where Truck = 1975 NHTSA truck definition), Annual Fuel1 Cost - Conventional Fuel, EPA Calculated Annual Fuel Cost - Conventional Fuel -----  Annual fuel cost error. Please revise Verify. , Intake Valves Per Cyl, Exhaust Valves Per Cyl, Carline Class, Car/Truck Category - Cash for Clunkers Bill., Unique Label?, Label Recalc?, Cyl Deact?, Var Valve Timing?, Var Valve Lift?, Fuel Metering Sys Cd, Camless Valvetrain (Y or N), Oil Viscosity, Stop/Start System (Engine Management System) Code, MFR Calculated Gas Guzzler MPG , FE Rating (1-10 rating on Label), GHG Rating (1-10 rating on Label), #1 Smog Rating Test Group, #1 Mfr Smog Rating (Mfr Smog 1-10 Rating on Label for

In [241]:
values = []
for i in data.loc[:,'$ You Save over 5 years (amount saved in fuel costs over 5 years - on label) ']:
    if i == 0:
        values.append('Not saved')
    else:
        values.append('Saved')

In [242]:
data["Amount saved on fuel over 5 years"] = values

In [243]:
data.drop(columns=['$ You Save over 5 years (amount saved in fuel costs over 5 years - on label) ','$ You Spend over 5 years (increased amount spent in fuel costs over 5 years - on label) '],axis=1,inplace=True)

In [244]:
data.shape

(1290, 41)

In [245]:
# Dividing columns into categorical and numerical
cat= [x for x in data.columns if data[x].dtypes=="O"]
num = [x for x in data.columns if data[x].dtypes!="O"]

In [246]:
data[cat].isnull().sum()

Division                                                          0  
Carline                                                           0  
Verify Mfr Cd                                                     0  
Air Aspir Method                                                  0  
Trans                                                             0  
Lockup Torque Converter                                           0  
Trans Creeper Gear                                                0  
Drive Sys                                                         0  
Fuel Usage  - Conventional Fuel                                   0  
Gas Guzzler Exempt (Where Truck = 1975 NHTSA truck definition)    0  
Car/Truck Category - Cash for Clunkers Bill.                      473
Unique Label?                                                     144
Label Recalc?                                                     62 
Cyl Deact?                                                        0  
Var Valve Timing?   

In [247]:
data['Car/Truck Category - Cash for Clunkers Bill.'].unique()

array(['car', nan, '1', '??'], dtype=object)

In [248]:
data['Unique Label?'].unique()

array(['N', 'Y', nan], dtype=object)

In [249]:
data['Label Recalc?'].unique()

array(['N', nan, 'Y'], dtype=object)

In [250]:
# Imputing null values with mode
def impute_nan_most_frequent_category(dataFrame,ColumnName):

     most_frequent_category=dataFrame[ColumnName].mode()[0]

     dataFrame[ColumnName + "_Imputed"] = dataFrame[ColumnName]
     dataFrame[ColumnName + "_Imputed"].fillna(most_frequent_category,inplace=True)

for Columns in ['Unique Label?','Label Recalc?']:
    impute_nan_most_frequent_category(data,Columns)

In [251]:
data.drop(columns = ['Unique Label?', 'Label Recalc?'],axis=1,inplace=True)

In [252]:
cat= [x for x in data.columns if data[x].dtypes=="O"]

In [253]:
sniff_modified(data[cat])

data type  \
Division                                                        object     
Amount saved on fuel over 5 years                               object     
Calc Approach Desc                                              object     
#1 Smog Rating Test Group                                       object     
Stop/Start System (Engine Management System) Code               object     
Oil Viscosity                                                   object     
Camless Valvetrain (Y or N)                                     object     
Fuel Metering Sys Cd                                            object     
Var Valve Lift?                                                 object     
Var Valve Timing?                                               object     
Unique Label?_Imputed                                           object     
Cyl Deact?                                                      object     
Gas Guzzler Exempt (Where Truck = 1975 NHTSA truck definition)  object     
Fuel Usage  - Conventional Fuel                                 object     
Drive Sys                                                       object     
Trans Creeper Gear                                              object     
Lockup Torque Converter                                         object     
Trans                                                           object     
Air Aspir Method                                                object     
Verify Mfr Cd                                                   object     
Carline                                                         object     
Car/Truck Category - Cash for Clunkers Bill.                    object     
Label Recalc?_Imputed                                           object     

                                                                percent missing  \
Division                                                        0.000000          
Amount saved on fuel over 5 years                               0.000000          
Calc Approach Desc                                              0.000000          
#1 Smog Rating Test Group                                       0.000000          
Stop/Start System (Engine Management System) Code               0.000000          
Oil Viscosity                                                   0.000000          
Camless Valvetrain (Y or N)                                     0.000000          
Fuel Metering Sys Cd                                            0.000000          
Var Valve Lift?                                                 0.000000          
Var Valve Timing?                                               0.000000          
Unique Label?_Imputed                                           0.000000          
Cyl Deact?                                                      0.000000          
Gas Guzzler Exempt (Where Truck = 1975 NHTSA truck definition)  0.000000          
Fuel Usage  - Conventional Fuel                                 0.000000          
Drive Sys                                                       0.000000          
Trans Creeper Gear                                              0.000000          
Lockup Torque Converter                                         0.000000          
Trans                                                           0.000000          
Air Aspir Method                                                0.000000          
Verify Mfr Cd                                                   0.000000          
Carline                                                         0.000000          
Car/Truck Category - Cash for Clunkers Bill.                    36.666667         
Label Recalc?_Imputed                                           0.000000          

                                                                No. unique  \
Division                                                        47           
Amount saved on fuel over 5 years                               2            
Calc Approa

In [254]:
# Dropping columns which are unrelated
data.drop(columns=['#1 Smog Rating Test Group','Division','Fuel Metering Sys Cd','Car/Truck Category - Cash for Clunkers Bill.'],axis=1,inplace=True)

In [255]:
# Encoding all other categorical columns
from sklearn import preprocessing
le = preprocessing.LabelEncoder()
data['Amount saved on fuel over 5 years'] = le.fit_transform(data['Amount saved on fuel over 5 years'])
data['Label Recalc?_Imputed'] = le.fit_transform(data['Label Recalc?_Imputed'])
data['Lockup Torque Converter'] = le.fit_transform(data['Lockup Torque Converter'])
data['Trans Creeper Gear'] = le.fit_transform(data['Trans Creeper Gear'])
data['Gas Guzzler Exempt (Where Truck = 1975 NHTSA truck definition)'] = le.fit_transform(data['Gas Guzzler Exempt (Where Truck = 1975 NHTSA truck definition)'])
data['Cyl Deact?'] = le.fit_transform(data['Cyl Deact?'])
data['Unique Label?_Imputed'] = le.fit_transform(data['Unique Label?_Imputed'])
data['Var Valve Timing?'] = le.fit_transform(data['Var Valve Timing?'])
data['Var Valve Lift?'] = le.fit_transform(data['Var Valve Lift?'])
data['Camless Valvetrain (Y or N)'] = le.fit_transform(data['Camless Valvetrain (Y or N)'])
data['Stop/Start System (Engine Management System) Code'] = le.fit_transform(data['Stop/Start System (Engine Management System) Code'])
data['Carline'] = le.fit_transform(data['Carline'])
data['Verify Mfr Cd'] = le.fit_transform(data['Verify Mfr Cd'])
data['Trans'] = le.fit_transform(data['Trans'])
data['Air Aspir Method'] = le.fit_transform(data['Air Aspir Method'])
data['Drive Sys'] = le.fit_transform(data['Drive Sys'])
data['Fuel Usage  - Conventional Fuel'] = le.fit_transform(data['Fuel Usage  - Conventional Fuel'])
data['Oil Viscosity'] = le.fit_transform(data['Oil Viscosity'])
data['Calc Approach Desc'] = le.fit_transform(data['Calc Approach Desc'])



In [256]:
data[num].isnull().sum()

Index (Model Type Index)                                                                                     0  
Eng Displ                                                                                                    0  
# Cyl                                                                                                        0  
Comb Unadj FE - Conventional Fuel                                                                            0  
# Gears                                                                                                      0  
Max Ethanol % - Gasoline                                                                                     0  
Max Biodiesel %                                                                                              0  
Annual Fuel1 Cost - Conventional Fuel                                                                        0  
EPA Calculated Annual Fuel Cost - Conventional Fuel -----  Annual fuel cost error. Please revise

In [257]:
# Checking for outliers
for k, v in data[num].items():
        q1 = v.quantile(0.25)
        q3 = v.quantile(0.75)
        iqr = q3 - q1
        v_col = v[(v <= q1 - 1.5 * iqr) | (v >= q3 + 1.5 * iqr)]
        perc = np.shape(v_col)[0] * 100.0 / np.shape(data)[0]
        print("Column %s's outliers =%d (%.2f%%)" % (k, np.shape(v_col)[0], perc))

Column Index (Model Type Index)'s outliers =2 (0.16%)
Column Eng Displ's outliers =61 (4.73%)
Column # Cyl's outliers =36 (2.79%)
Column Comb Unadj FE - Conventional Fuel's outliers =27 (2.09%)
Column # Gears's outliers =40 (3.10%)
Column Max Ethanol % - Gasoline's outliers =92 (7.13%)
Column Max Biodiesel %'s outliers =1290 (100.00%)
Column Annual Fuel1 Cost - Conventional Fuel's outliers =49 (3.80%)
Column EPA Calculated Annual Fuel Cost - Conventional Fuel -----  Annual fuel cost error. Please revise Verify. 's outliers =49 (3.80%)
Column Intake Valves Per Cyl's outliers =1290 (100.00%)
Column Exhaust Valves Per Cyl's outliers =1290 (100.00%)
Column Carline Class's outliers =0 (0.00%)
Column MFR Calculated Gas Guzzler MPG 's outliers =22 (1.71%)
Column FE Rating (1-10 rating on Label)'s outliers =68 (5.27%)
Column GHG Rating (1-10 rating on Label)'s outliers =57 (4.42%)
Column #1 Mfr Smog Rating (Mfr Smog 1-10 Rating on Label for Test Group 1)'s outliers =0 (0.00%)
Column City CO2 R

In [258]:
# Imputing mean into null values
from sklearn.impute import SimpleImputer
feature = ['MFR Calculated Gas Guzzler MPG ','EPA Calculated Annual Fuel Cost - Conventional Fuel -----  Annual fuel cost error. Please revise Verify. ']
imputer = SimpleImputer(missing_values=np.nan, strategy='mean')
imputer = imputer.fit(data[feature])
data[feature] = imputer.transform(data[feature])

In [259]:

data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1290 entries, 0 to 1289
Data columns (total 37 columns):
 #   Column                                                                                                     Non-Null Count  Dtype  
---  ------                                                                                                     --------------  -----  
 0   Carline                                                                                                    1290 non-null   int32  
 1   Verify Mfr Cd                                                                                              1290 non-null   int32  
 2   Index (Model Type Index)                                                                                   1290 non-null   int64  
 3   Eng Displ                                                                                                  1290 non-null   float64
 4   # Cyl                                                                   

In [262]:
cor = data.corr()
cor.sort_values([ 'Comb Unadj FE - Conventional Fuel'], ascending=False, inplace=True)
cor[ 'Comb Unadj FE - Conventional Fuel']

Comb Unadj FE - Conventional Fuel                                                                            1.000000
FE Rating (1-10 rating on Label)                                                                             0.965413
GHG Rating (1-10 rating on Label)                                                                            0.959237
MFR Calculated Gas Guzzler MPG                                                                               0.891531
Amount saved on fuel over 5 years                                                                            0.709165
#1 Mfr Smog Rating (Mfr Smog 1-10 Rating on Label for Test Group 1)                                          0.361955
Exhaust Valves Per Cyl                                                                                       0.282873
Intake Valves Per Cyl                                                                                        0.261015
Trans                                                   

### Part 2: Documentation

In this part you should document everything that you did to prepare the data to be used to train a random forest model. 

**Note:** This part does not need to include any code and should be formatted so it's easy to read. See 'Help -> Markdown Reference' in Jupyter Lab menu at top left of the screen. 

### Study the data
- Using head, shape, and sniff modified function we studied the dataset. We learned that there are 1290 samples of different vehicles and 164 features.
- Using sniff modified function we could extract null values, count of unique value and unique values in the columns.

### Transforming the data
- Now we have dropped around 20 columns with 100% null values which are of no use for the model.
- As per the instructions we have dropped columns related to fuel economy except our target variable which is 'comb unadj FE'.
- Next we have dropped description columns as they are of no use to the target. Moreover they are just description of other columns. For example there are columns called 'Air Aspir Method' which have values of different aspiration methods in short form and 'Air Aspir Method Desc' has the long form of the methods mentioned in the earlier columns.
- Next we have dropped Passenger and luggage volume columns. Though they are important to target variable, these columns have more than 60% null values and there is no logical way to impute them. Because there are columns where all 2WD pass & lugg, 4WD pass & lugg, and Hatchback pass & lugg are null values. if that was not the case we could have assumed that 2WD pass & lugg values are nulls because they are either 4WD or hatchback vehicles.
- Next we dropped columns related to batteries, regen, and fuel2 because there are around 90% null values in these columns and there is no possible way to impute the nulls. Moreover fuel2 columns which are columns related to alternate fuel are dropped because we are trying find out the fuel economy of conventional fuel not alternate fuel.
- Guzzler is a component in automobile which causes heavy fuel consumption and related to fuel economy. But there are more than 90% null values and cannot logically impute them.
- Range1 - Model Type Driving Range - Conventional Fuel, Relabel, # Drive Motor Gen','Motor Gen Type, If Other','Rated Motor Gen Power (kW)','Fuel Cell Vehicle (Y or N)','Off Board Charge Capable (Y or N)','GHG 1-10 rating on Ethanol (EPA Determined)','#2 Smog Rating Test Group','#2 Mfr Smog Rating (Mfr Smog 1-10 Rating on Label for Test Group 2)','Unnamed: 0'
- These columns are either unrelated like fuel unit or have more null values which cannot be replaced logically.
- Transmission column is removed because there are columns called Trans and Gears which says the same information. For example there is a value in Transmission which says Auto(S8) which means semi automatic vehicle with 8 gears. If we compare with Trans and Gears, the transmission values is divided as SA in Trans and 8 in Gears.
- Mfr Name is dropped because there is other column called verify mfr code which is a short of the Mfr name.

- Model Year, Suppressed, and Police/Emerg are dropped because they have only one unique value.

- Earlier we have dropped description columns because every description column has the parent column, but Calc Approach Desc has not parent column so we again added that column. Otherwise it will cause loss of information.

- In Air Aspir Method there is a value called Naturally Aspirated(from its description column) and this value is written in short form 'NA'. And this NA values are considered as null values in python. So we replaced the nulls with NAT.

- Max Ethanol and Max Biodiesel are linked with type of fuel used in the vehicles.  Gasoline has the chemical component ethanol and diesel has the chemical component biodiesel. We have checked that where ever there are null values in ethanol the fuel used in diesel and biodiesel has the value. So we replaced nulls with 0 in both ethanol and biodiesel.

- There are 2 columns which says how much the person saved on fuel in last 5 years and how much the person spent extra on fuel in last 5 years. So here we have seen that where ever there are nulls in "$ You Save over 5 years (amount saved in fuel costs over 5 years - on label)', there are values in '$ You Spend over 5 years (increased amount spent in fuel costs over 5 years - on label) '. So we replaced the nulls in first column with 0 and created a new column called whether the amount is saved or not. If there are 0 in $ You Save over 5 years (amount saved in fuel costs over 5 years - on label) then not saved else saved. And we dropped the original columns.

- Next we seperated categorical and numerical variables. In category variables the column 'Car/Truck Category - Cash for Clunkers Bill.' Have less null values we cannot replace them so we dropped it.

- Using categorical imputation method we have imputed nulls in 'Unique Label?' and 'Label Recalc?' with mode of their respective columns.

- Using Label encoding we have encoded categorical variables. This method converts values in categorical in numbers from 0.

- In numeric variables, first we checked for outliers. Then using imputer function from sklearn we have replaced nulls with their mean. Because there are not much outliers in these columns.
- Finally we have got the dataset of 1290*41 after the data preprocessing. And everything is in numerical data type. 